### PR TITLE
Update attention.cpp to remove warning about preferring torch.bool type

### LINKED ISSUE #100469: TransformerEncoderLayer always warns when using src_key_padding_mask in inference mode.
### 🐛 Describe the bug

```python

import torch

with torch.inference_mode():
    d_model = 4
    layer = torch.nn.TransformerEncoderLayer(d_model, 2, 2, batch_first=True)
    layer.eval()
    x = torch.randn(5, 10, d_model)
    pad = torch.rand(5, 10) > 0.5
    layer(x, src_key_padding_mask=pad)
```

Gives the warning:
```
/home/jhseu/.local/lib/python3.10/site-packages/torch/nn/modules/transformer.py:668: UserWarning: Converting mask without torch.bool dtype to bool; this will negatively affect performance. Prefer to use a boolean mask directly. (Triggered internally at ../aten/src/ATen/native/transformers/attention.cpp:152.)
  return torch._transformer_encoder_layer_fwd(
```

This is because the layer is converting it to a float mask regardless of what the user passes, and then ATen warns about requiring a bool mask. This warning appears on every call when doing inference with libtorch.

### Versions

Collecting environment information...
PyTorch version: 2.0.0+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 12.2.1 20230201
Clang version: 15.0.7
CMake version: version 3.25.0
Libc version: glibc-2.37

Python version: 3.10.10 (main, Mar  5 2023, 22:26:53) [GCC 12.2.1 20230201] (64-bit runtime)
Python platform: Linux-6.2.13-arch1-1-x86_64-with-glibc2.37
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 530.41.03
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.8.0
/usr/lib/libcudnn_adv_infer.so.8.8.0
/usr/lib/libcudnn_adv_train.so.8.8.0
/usr/lib/libcudnn_cnn_infer.so.8.8.0
/usr/lib/libcudnn_cnn_train.so.8.8.0
/usr/lib/libcudnn_ops_infer.so.8.8.0
/usr/lib/libcudnn_ops_train.so.8.8.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   43 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          32
On-line CPU(s) list:             0-31
Vendor ID:                       AuthenticAMD
Model name:                      AMD Ryzen Threadripper 1950X 16-Core Processor
CPU family:                      23
Model:                           1
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        1
Frequency boost:                 enabled
CPU(s) scaling MHz:              68%
CPU max MHz:                     3400.0000
CPU min MHz:                     2200.0000
BogoMIPS:                        6789.07
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid amd_dcm aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate ssbd vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca sev
Virtualization:                  AMD-V
L1d cache:                       512 KiB (16 instances)
L1i cache:                       1 MiB (16 instances)
L2 cache:                        8 MiB (16 instances)
L3 cache:                        32 MiB (4 instances)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-31
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; untrained return thunk; SMT vulnerable
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.1
[pip3] pytorch-lightning==2.0.1
[pip3] torch==2.0.0+cu118
[pip3] torchaudio==2.0.1+cu118
[pip3] torchmetrics==0.11.4
[pip3] torchvision==0.15.1+cu118
[pip3] triton==2.0.0
[conda] Could not collect

cc @jbschlosser @albanD @mruberry @walterddr @mikaylagawarecki @bhosmer @cpuhrsch @erichan1 @drisspg

### PR DESCRIPTION
**Pull request by mikekgfb**
Update attention.cpp to remove warning about preferring torch.bool data type 

Fixes #100469 #97532

### CODE REVIEW

**Comment by mikekgfb** [comment]
Masks can be floating point (and in fact are preferred to be floating point for SDPA), so for consistency we should allow float masks for creating Nested Tensors from vanilla (dense) tensors in conjunction with a float-typed padding mask with the convention of -infinity being a mask-out

https://github.com/pytorch/pytorch/blob/03101a227f6639d5a9ad628d1dc300f9f99a8812/aten/src/ATen/native/nested/NestedTensorMath.cpp#L63
https://github.com/pytorch/pytorch/blob/03101a227f6639d5a9ad628d1dc300f9f99a8812/aten/src/ATen/native/nested/NestedTensorMath.cpp#L94

**Review by mikaylagawarecki — APPROVED** [review]

**Review by mikaylagawarecki — CHANGES_REQUESTED** [review]
Sorry, from offline discussion with @jbschlosser, removing the warning in masked_softmax is probably ok but perhaps we should not be converting the float masks to bool masks in the NestedTensor_from_mask -- the mask seems to be getting converted from bool-->float-->bool again (perhaps the fix could be localized to[`_canonical_mask` where we are doing the conversion from bool --> float once](https://github.com/pytorch/pytorch/blob/main/torch/nn/functional.py#L5057))

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 1]
```diff
@@ -60,15 +60,23 @@ Tensor pad_tensor_to_shape(
...
     TORCH_CHECK(t.dim() == 3, "Input should be a 3D tensor, N * L * D");
     auto N = t.size(0), L = t.size(1), D = t.size(2);
     auto NN = mask.size(0), LL = mask.size(1);
     TORCH_CHECK(N == NN && L == LL, "Mask size should match input size");
 
     // N * L
-    Tensor sizes = mask;
+    Tensor sizes;
+    if (mask.scalar_type() != at::ScalarType::Bool){
+      sizes = mask.to(at::kBool);
```
Comment (jbschlosser) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:75]: Is my understanding here correct? If the user passes a bool mask into `TransformerEncoderLayer`, it will be converted to a float in `F._canonical_mask()`. Then it will be reconverted back to bool here.

If so, I'd prefer we avoid the conversion gymnastics. Can we just change `F._canonical_mask()` somehow to do it right in the first place?

**Comment by mikaylagawarecki** [comment]
cc @i-markov

File: aten/src/ATen/native/transformers/attention.cpp [hunk 2]
```diff
@@ -147,9 +147,6 @@ Tensor masked_softmax(
     return at::_nested_tensor_softmax_with_shape(attn_scores, query);
   }
   if (attn_mask && attn_mask->dtype() != at::kBool) {
-    TORCH_WARN(
```
Comment (jbschlosser) [inline aten/src/ATen/native/transformers/attention.cpp:150]: I think there should be a way to avoid triggering this warning without actually removing it from the codebase.

**Review by jbschlosser — CHANGES_REQUESTED** [review]
I have some concerns about the mask conversions and am hoping we can take a simpler approach here.

Disclaimer: I haven't had a ton of time to dig into this in detail.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp (hunk 1 above)
Comment (mikekgfb) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:75]: Joel, a few considerations, and I'm happy to do this another way:

1 - if we say a mask can be floating point (which we say pervasively everywhere), then we shoulkd accept it here.  
2 - you are right, in this scenario for the fastpath there are two conversions. It's not so easy though...

Considerations:
* we need a very detailed set of conditions to find whether this should be float or bool.  Because when it goes to sdpa, float is the preferred way, otherwise we have a conversion in every layer.
* so, we'd have to duplicate the analysis before canonical mask -- or sink canonical mask further down.  That's feasible (at some logic complexity, like we need a "check_only" boolean, and the path will be more interesting.  But manageable
* switching to float not only avoided the conversions on the sdpa, path, it also allowed us to handle mask logic more easily.  Because masks must be folded at various points, and that uses to check wthe data type and eiother add the masks (for float masks) or do logicals (for bool masks).  So we have to add the bool mask merge logic back everywhere.
* double conversion bool-float-bool happens only once on the way to Nested Tensor creation (once for all layers), bool float conversion if a bool is passed to sdpa happens in every layer.  
* we are talking about refactoring the fastpath to take advantage. of PT2 compile (since the current Nested Tensors are not supproted by `torch.compile()`.  So just wanted to switch off annoying warnings reported by our users (and the warnings are frankly incorrect, because even if they pass in bool we convert it to float and warn the users to please pass in bool.  Ummm...)  and improve the general situation (such as allowing float masks in nested tensor creation).

This is not a perfect solution, I did feel it made things notably better.  In my mind a perfect solution is more expensive, and actually not that perfect, and not again, the refactor just screamed make it better with limited engineering.  Doing nothing feels like less than this.

File: aten/src/ATen/native/transformers/attention.cpp (hunk 2 above)
Comment (mikekgfb) [reply aten/src/ATen/native/transformers/attention.cpp:150]: Supposedly not so well, and there were particular complaints about Torchscript.  But if you can suggest a solution.  
That being said, even then, the warning is sorta incorrect, because we end up warning the user even if she passed in a bool, because we converted it for them.)

**Comment by mikekgfb** [comment]
Updated TransformerEncoder to keep the original input for key src padding mask and mask.  When we go to the C++ fastpath, we use the original inputs to avoid double conversions.

I suggest we keep the framework for supporting floats more efficiently by recognizing floating point masks properly for fastpath (without warnings) because for all other execution modes we actually prefer float so we shouldn't aggressively encourage users to use bool instead of float.

**Comment by mikekgfb** [comment]
@mikaylagawarecki @jbschlosser can you please have a look to see whether this addresses Joel's comments about double conversion?

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 3]
```diff
@@ -60,15 +60,23 @@ Tensor pad_tensor_to_shape(
...
     TORCH_CHECK(mask.dim() == 2, "Padding mask should be 2D");
     TORCH_CHECK(t.dim() == 3, "Input should be a 3D tensor, N * L * D");
     auto N = t.size(0), L = t.size(1), D = t.size(2);
     auto NN = mask.size(0), LL = mask.size(1);
     TORCH_CHECK(N == NN && L == LL, "Mask size should match input size");
 
     // N * L
-    Tensor sizes = mask;
+    Tensor sizes;
+    if (mask.scalar_type() != at::ScalarType::Bool){
```
Comment (mikaylagawarecki) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:74]: hm, why do we still need to allow floats here and in `NestedTensor_nested_tensor_from_mask_left_aligned` if we do `src_key_padding_mask_.logical_not()` when passing the mask argument in which iiuc will return a bool mask below?

Comment (mikekgfb) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp]: Removed.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 4]
```diff
@@ -61,6 +61,9 @@ Tensor pad_tensor_to_shape(
 
 Tensor NestedTensor_nested_tensor_from_mask(const Tensor& t, const Tensor& mask, bool mask_check) {
     TORCH_CHECK(mask.scalar_type() == at::ScalarType::Bool, "Expected mask to be of ScalarType Bool, but got ", mask.scalar_type(), " instead.");
```
Comment (mikaylagawarecki) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:63]: seems like this line should be removed :)

File: aten/src/ATen/native/transformers/attention.cpp (hunk 2 above)
Comment (mikaylagawarecki) [reply aten/src/ATen/native/transformers/attention.cpp:150]: For posterity I think this is ok because I believe the only non-transformers usecase for `masked_softmax` where this warning might be useful is MaskedTensor

File: torch/nn/modules/transformer.py [hunk 5]
```diff
@@ -273,10 +278,13 @@ def forward(
...
             why_not_sparsity_fast_path = f"input not batched; expected src.dim() of 3 but got {src.dim()}"
-        elif src_key_padding_mask is None:
+        # src_key_padding_mask_ and src_key_padding_mask are the same, minus type
+        # conversion. Chacking <var>_ version for torchscript type refinement
+        # but reporting named input in performance debug hint
+        elif src_key_padding_mask_ is None:
             why_not_sparsity_fast_path = "src_key_padding_mask was None"
         elif (((not hasattr(self, "mask_check")) or self.mask_check)
-                and not torch._nested_tensor_from_mask_left_aligned(src, src_key_padding_mask.logical_not())):
+                and not torch._nested_tensor_from_mask_left_aligned(src, src_key_padding_mask_.logical_not())):
```
Comment (mikaylagawarecki) [inline torch/nn/modules/transformer.py:287]: I'm sorry I didn't catch this in the previous pass but reading this again I am actually somewhat confused here, it seems like we are doing 

1) save `src_key_padding_mask` as `src_key_padding_mask_`
2)`src_key_padding_mask =  _canoncial_mask(...)` which will do bool --> float conversion
3) call `src_key_padding_mask.logical_not()` which will make the mask passed in a boolean mask again

So do we need to save `src_key_padding_mask_` at all? Similarly for `mask`

File: aten/src/ATen/native/nested/NestedTensorMath.cpp (hunk 4 above)
Comment (mikekgfb) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:63]: That's what happens when I try to do things quick.  I agree that we're not calling this today.  So, we don't need to update the check.  (And the change does exactly that, because I added the old check back but had not taken the second check out.  )

The original change was not motivated by us calling this function with float today, but rather with the intent that "we're treating floats as legal type for masks, so let's make them legal wherever masks go".  Maybe a bit over-eager ;)
So, removing that bit and dialing it back to where it was

**Review by mikaylagawarecki — APPROVED** [review]
Thanks!
